src/app/api/game-sessions/route.ts

```diff
@@ -20,21 +91,280 @@ export async function GET() {
+      return new Response("Start time must be before end time", {
+        status: 400,
+      });
+    }
+
+    if (gameSessionToInsert.casualCapacity > gameSessionToInsert.capacity) {
+      return new Response("Casual capacity must be less than capacity", {
+        status: 400,
+      });
+    }
+
+    await db.insert(gameSessions).values(gameSessionToInsert);
+
+    return NextResponse.json(gameSessionToInsert, { status: 201 });
   } catch (error) {
     if (error instanceof z.ZodError)
       return NextResponse.json(error.issues, { status: 400 });
 
     return new Response(null, { status: 500 });
   }
 }
+
+export async function DELETE(req: NextRequest) {
+  try {
+    const user = await getCurrentUser();
+    if (!user) {
+      return new Response("ERROR: Unauthorized request", { status: 401 });
+    }
+    if (user.role != "admin") {
+      return new Response("ERROR: No valid permissions", { status: 403 });
+    }
+
+    const date = req.nextUrl.searchParams.get("date");
+    const gameSessionDate = z.string().date().parse(date);
+
+    // Delete the game session if it exists
+    const [deletedGameSession] = await db
+      .delete(gameSessions)
+      .where(eq(gameSessions.date, gameSessionDate))
+      .returning();
+
+    if (!deletedGameSession) {
+      // Create a gameSessionException record if one doesn't exist
+      const gameSessionException =
+        await db.query.gameSessionExceptions.findFirst({
+          where: eq(gameSessionExceptions.gameSessionDate, gameSessionDate),
+        });
+
+      if (gameSessionException) {
+        return new Response(
+          "A gameSessionException already exists for this date",
+          { status: 404 },
+        );
+      } else {
+        const gameSessionExceptionToInsert =
+          insertGameSessionExceptionSchema.parse({
+            gameSessionDate,
+          });
+
+        await db
+          .insert(gameSessionExceptions)
+          .values(gameSessionExceptionToInsert);
+      }
+    }
+
+    return new Response(null, { status: 204 });
+  } catch (error) {
+    if (error instanceof z.ZodError) {
+      return NextResponse.json(error.issues, { status: 400 });
+    }
+    return new Response("Internal Server Error", { status: 500 });
+  }
+}
+
+export async function PUT(req: NextRequest) {
+  try {
+    const user = await getCurrentUser();
+    if (!user) {
+      return new Response("ERROR: Unauthorized request", { status: 401 });
+    }
+    if (user.role != "admin") {
+      return new Response("ERROR: No valid permissions", { status: 403 });
+    }
+
+    const date = req.nextUrl.searchParams.get("date");
+    const gameSessionDate = z.string().date().parse(date);
+
+    const body = await req.json();
+
+    const gameSessionToUpdate = updateGameSessionSchema.parse(body);
+    const gameSessionToInsert = insertGameSessionSchema.parse(body);
+
+    if (gameSessionToInsert.date !== gameSessionDate) {
```

Comment:
should this be gameSessionToUpdate

Reply:
No gameSessionToUpdate does not have a date property so we have to check gameSessionToInsert. I have also reordered this to make it more clear when I moved the other checks to earlier in the function